File: 05월 24일 - 유니온 파인드/과제/20040.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> parent;

//Find 연산
int findParent(int node) {
    if (parent[node] < 0) {
        return node;
    }
    return parent[node] = findParent(parent[node]);
}

//Union 연산
bool unionInput(int x, int y) {
    int xp = findParent(x);
    int yp = findParent(y);

    if (xp == yp) {
        return false;
    }
    if (parent[xp] < parent[yp]) {
        parent[xp] += parent[yp];
        parent[yp] = xp;
    } else {
        parent[yp] += parent[xp];
        parent[xp] = yp;
    }
    return true;
}

/**
 * [사이클 게임]
 *
 * 사이클이 발생한 순간 = 같은 집합에 있는 원소 두 개를 유니온하려 할 때
 * unionInput 함수의 반환형을 bool로 선언하여 cycle이 생성되는 순간 발견하기
 */

int main() {
    int n, m, x, y;

    //입력
    cin >> n >> m;
    parent.assign(n, -1);
    for (int i = 0; i < m; i++) {
        cin >> x >> y;

        //연산 & 출력
        if (!unionInput(x, y)) { //사이클이 생성됨
            cout << i + 1;
            return 0;
        }
    }
    cout << 0;
    return 0;
}

File: 05월 24일 - 유니온 파인드/과제/20040.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

"""
 [사이클 게임]

 사이클이 발생한 순간 = 같은 집합에 있는 원소 두 개를 유니온하려 할 때
 union 함수의 리턴값을 boolean으로 주어 cycle이 생성되는 순간 발견하기
"""

# find 연산
def find_parent(x):
    if parent[x] < 0:
        return x
    parent[x] = find_parent(parent[x])
    return parent[x]

# union 연산
def union(x, y):
    px = find_parent(x)
    py = find_parent(y)

    if px == py:
        return False

    if parent[px] < parent[py]:
        parent[px] += parent[py]
        parent[py] = px
    else:
        parent[py] += parent[px]
        parent[px] = py

    return True

# 입력
n, m = map(int, input().split())
parent = [-1]*n     # 초기화

for i in range(m):
    x, y = map(int, input().split())
    if not union(x, y):
        # 사이클이 생성됨
        print(i+1)
        break
else:   # for-else문: for문이 break에 걸리지 않고 정상 종료된 경우에만 else문 실행
    print(0)